IBX-5609 - Add Extend Payment article (#1996)
ezsystems/developer-documentation: docs/commerce/payment/extend_payment.md

---
description: Extend Payment with custom payment method type.
edition: commerce
---

# Extend Payment

There are different ways you can extend your Payment module implementation. 
One of them is to create a custom payment method type. 
The other is attaching custom data to a payment.

You can also [customize the payment processing workflow](configure_payment.md#custom_payment_workflows).

## Create custom payment method type

> If your application needs payment methods of other type than the default `offline` one, you must create custom payment method types. 

Reviewer: Maybe add a sentence t the tune of: the example below shows this on the example of PayPal...

!!! note "Gateway integration requirement"

>     [[= product_name =]] does not come with gateway redirects. Whether you are an integrator or the end customer, it is your responsibility to implement payment gateway integration.

Reviewer: ```suggestion
    [[= product_name =]] does not come with gateway redirects. Whether you are an integrator or an end customer, it is your responsibility to implement payment gateway integration.
```

### Define custom payment method type

Create a PHP definition of the payment method type.

``` php
[[= include_file('code_samples/front/shop/payment/src/PaymentMethodType/PayPal/PayPal.php') =]]
```

> Make sure that `getName` returns a human readable name of the payment method type, the way you want it to appear on the list of available payment method types.

Reviewer: ```suggestion
Make sure that `getName()` returns a human-readable name of the payment method type, the way you want it to appear on the list of available payment method types.
```

Now, register the definition as a service:

``` yaml
[[= include_file('code_samples/front/shop/payment/config/packages/services.yaml', 0, 5) =]]

> 

Reviewer: ```suggestion
```

```

> As an alternative, instead of creating a PHP library, you can use the Type Factory to define the payment method type in your YAML configuration file:

Reviewer: ```suggestion
As an alternative, instead of creating a custom class, you can use a built-in type factory to define the payment method type in the service definition file:
```

``` yaml
services:
    app.payment.type.paypal: 
        class: Ibexa\Contracts\Payment\PaymentMethod\Type\TypeInterface
        factory: [ '@Ibexa\Contracts\Payment\PaymentMethod\Type\TypeFactoryInterface', 'createType' ]
        arguments: 
            $identifier: 'paypal' 
            $name: 'PayPal' 
        tags: 
            - name: ibexa.payment.payment_method.type 
              alias: paypal

```

> At this point a custom payment method type should be visible in the user interface.

Reviewer: Will it be visible at this point, before you add forms? Or only visible in the payment list, but not when you click it?

### Create options form

Create a corresponding form type:

``` php
[[= include_file('code_samples/front/shop/payment/src/Form/Type/PayPalOptionsType.php') =]]
```

Next, create a mapper that maps the information that the user inputs in the form into attribute definition.

``` php
[[= include_file('code_samples/front/shop/payment/src/PaymentMethodType/PayPal/OptionsFormMapper.php') =]]
```

> Then, register the `OptionsFormMapper` interface as a service:

Reviewer: ```suggestion
Then, register `OptionsFormMapper` as a service:
```

``` yaml
services:
    App\Payment\PayPal\OptionsFormMapper:
        tags:
            -   name: ibexa.payment.payment_method.options.form_mapper
                type: paypal
```

### Create options validator

You might want to make sure that data provided by the user is validated. 
To do that, create an options validator that checks user input against the constraints and dispatches an error when needed.

``` php
[[= include_file('code_samples/front/shop/payment/src/PaymentMethodType/PayPal/UrlOptionValidator.php') =]]
```

Then, register the validator as a service:

``` yaml
services:
    App\Payment\PayPal\OptionsValidator:
        tags:
            -   name: ibexa.payment.payment_method.options.validator
                type: paypal
```

### Restart application

> Restart the application and create a payment of the new type.

Reviewer 2: What do you mean by "Restart application" ?

Author: It is a term we've used in other places of the doc, telling people to shut down and start the server again.

> ![Payment method of custom type](img/custom_paymant_type.png "Payment method of custom type")

Reviewer: ```suggestion
![Payment method of custom type](custom_paymant_type.png "Payment method of custom type")
```
